Added a config struct for the dvl_teledyne driver

JanWehrmann opened the pull request:
and an interface to send those settings to the device.

saarnold commented on src/Driver.cpp:
```diff
@@ -184,3 +165,228 @@ void Driver::startAcquisition()
+    setDepthCellSize(conf.depthCellSize);
+
+    // Save in case communication fails after altering serial port control settings.
+    writePacket(reinterpret_cast<uint8_t const*>("CK\n"), 3, 100);
```
Here is a readConfigurationAck missing.

saarnold commented on src/Driver.cpp:
```diff
@@ -184,3 +165,228 @@ void Driver::startAcquisition()
+    return numDigits;
+}
+
+std::vector<int> Driver::splitTime(const base::Time& microseconds) const
```
You could use time.toString() and then extract the values from there.

doudou replied:
:-1: there's no guarantee on the format of toString. What he does is the right thing and/or he could have also added the corresponding methods to base::Time itself.

saarnold replied:
Right, it takes the local timezone into account, which is undesired here.
Edit: And in general only shows the local time, not the absolute time in e.g. hours.
But why is the format not guaranteed?

@JanWehrmann 
Feel free to move the method to `base::Time` and create a PR there.

doudou replied:
> But why is the format not guaranteed?

As in "if someone wants to change the format of toString, there is nothing forbidding him to do so". There is no canonical way to display time as a string. For instance, at some point, somebody added milliseconds (and/or microseconds, as far as I remember).

doudou replied:
Generally speaking, don't use a conversion to string to convert what's essentially a numerical value to another numerical value. Conversions to string are usually meant for user consumption, not data transfer.

2maz replied:
You might miss here, that you can use toString (and fromString) with arguments to specify the format you generate/read from?

```
 std::string toString(base::Time::Resolution resolution = Microseconds, const std::string& mainFormat = "%Y%m%d-%H:%M:%S") const;

doudou replied:
I did miss that, yes. The "format is not consistent" argument is therefore moot. The "it's seriously ugly" argument still stands though.

saarnold commented on src/Driver.cpp:
```diff
@@ -184,3 +165,228 @@ void Driver::startAcquisition()
+    for (int i = 0; i < 4; ++i)
+    {
+        std::string timeString;
+        if (getNumberOfDigits(i) == 1)
```
You probably intended to check `getNumberOfDigits(itimeSplits.at(i)) == 1` here right?

saarnold commented on src/Driver.cpp:
```diff
@@ -184,3 +165,228 @@ void Driver::startAcquisition()
+    readConfigurationAck();
+}
+
+int Driver::getNumberOfDigits(int value) const
```
What about converting the integers to string before and then just checking the size of the resulting string?
E.g. `std::to_string(my_int).size()`

JanWehrmann replied:
Good idea. I will remove the function and just check the stringsize where needed.

saarnold commented:
Depends on rock-core/base-types#122

doudou commented on CMakeLists.txt:
```diff
@@ -3,4 +3,7 @@
 find_package(Rock)
 rock_init(dvl_teledyne 0.1)
+
+SET( CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS} -std=c++0x" )
```
Use CMake's support for C++11/C++14 please. See https://rock-core.github.io/rock-and-syskit/integrating_functionality/cpp_libraries.html#cxx_standard

and in base/cmake: https://github.com/rock-core/base-cmake/blob/master/modules/Rock.cmake#L9

JanWehrmann replied:
Replaced with 
set(CMAKE_CXX_STANDARD 11)

doudou commented on src/Config.hpp:
```diff
@@ -0,0 +1,278 @@
+struct Config
+{
+    Config()
+    : bottomTrackPingsPerEnsemble(1)
```
Just a recommendation, but since you're switching the package to C++11, it would IMO be nicer to use the field initialization syntax, e.g.

~~~
int maximumTrackingDepth = 100;
~~~

doudou commented on src/Driver.hpp:
```diff
@@ -63,6 +61,79 @@ namespace dvl_teledyne
+        void setHeadingAlignment(float headingAlignment);
+
+        /** Same as setHeadingAlignment but corrects for electrical and magnetic bias
+         * instead of physikal misalignment.*/
```
`physical`

doudou commented on src/Driver.hpp:
```diff
@@ -63,6 +61,79 @@ namespace dvl_teledyne
+                                    bool enableDataRecording);
+
+        /** Configures the correction between the heading reference and beam 3 of the
+         * DVL (-17999 - 18000 hundreths of a degree).
```
The documentation is ambiguous ... one could think it applies to the function's argument.
```suggestion
         * DVL
         * @param headingAlignment offset in radians, between -179.99 and +180 degrees
```

Maybe use a `base::Angle` instead of a plain float ?

JanWehrmann replied:
Thanks! I forgot to update the documentation. I also adapted the rest of the documentation to this style where it seemed appropriate.

JanWehrmann replied:
Using base::Angle seems like a good idea though the documentation states it only holds values
 PI < value <= PI
so I think there is a negative sign missing.

doudou commented on src/Driver.hpp:
```diff
@@ -63,6 +61,79 @@ namespace dvl_teledyne
+        /** Configures the number of pings per data ensemble during water profiling (0 - 16384).*/
+        void setPingsPerEnsemble(int pingsPerEnsemble);
+
+        /** Configures the height of one measurement cell (10 - 800 cm).*/
```
Same comment than for setHeadingAlignment

```suggestion
        /** Configures the height of one measurement cell
         * @param depthCellSize the size of a depth cell in meters. Must be between 0.01 and 8.
         */
```

JanWehrmann replied:
Applied your suggestions.

saarnold commented on src/Driver.cpp:
```diff
@@ -178,9 +160,216 @@ void Driver::startAcquisition()
+    else if (sign)
+        command += "+";
+
+    std::string valueString = std::to_string(value);
```
In case of a negative value the command string would contain two negative signs.

JanWehrmann replied:
Good point. I used std::abs instead of removing the logic for setting minus signs, to preserve the padding logic later on.

saarnold commented on src/Driver.cpp:
```diff
@@ -178,9 +160,216 @@ void Driver::startAcquisition()
+    if (expert)
+        ++numDigits;
+
+    writePacket(reinterpret_cast<uint8_t const*>((command + "\n").c_str()), numDigits + 3);
```
The length of the resulting command should be equal to the amount of bytes you want to send.
And since the expert flag is related to the length of the initial command and not to the digits you shoudn't need it.

JanWehrmann replied:
You're right. Removed the flag.

JanWehrmann replied:
You're right. Removed the flag.

saarnold commented on src/Main.cpp:
```diff
@@ -17,7 +17,7 @@ int main(int argc, char const* argv[])
 
     dvl_teledyne::Driver driver;
-    driver.open(argv[1]);
+    driver.openURI(argv[1]);
```
You would need to start the acquisition now manually right?
Can you document the intended usage of this driver (open, configuration, start acquisition) in the header of the driver?

JanWehrmann replied:
Done.

saarnold replied:
You still would need to start the acquisition here in order to read samples right?
Same for the dvl_teledyne_info executable. And lets also stop the acquisition again in all executables.

JanWehrmann replied:
Also done.

saarnold commented on src/CMakeLists.txt:
```diff
@@ -12,3 +12,6 @@ rock_executable(dvl_teledyne_read
 rock_executable(dvl_teledyne_configure
     MainConfigure.cpp
     DEPS dvl_teledyne)
+rock_executable(dvl_teledyne_config_configure
```
Since this applies the default config, lets name it `dvl_teledyne_apply_default_config`?

JanWehrmann replied:
Sounds good!

JanWehrmann replied:
I also changed the name of the source file to MainApplyDefaultConfig.cpp

saarnold commented on src/Driver.cpp:
```diff
@@ -178,9 +142,208 @@ void Driver::startAcquisition()
+    for (int i = 1; i < 5; ++i)
+    {
+        if (i == 1)
+            timeValues.at(i) /= 10;
```
Since you want to do this only for one element why doing it inside the loop?

JanWehrmann replied:
Good point. Moved it in front of the loop.

JanWehrmann committed:
Added a config struct for the dvl_teledyne driver

and an interface to send those settings to the device.

saarnold commented on src/MainConfigure.cpp:
```diff
@@ -29,6 +29,18 @@ int main(int argc, char const* argv[])
     // configuration mode
     driver.setConfigurationMode();
     driver.startAcquisition();
+
```
Since this just applies a config file, why starting the acquisition here in the first place?

JanWehrmann replied:
Removed it.

saarnold commented:
There are currently commits in this PR that are already on master. My fault, the target branch needs to be rebased.

JanWehrmann committed:
Removed "open" and "setDesiredBaudrate" methods.

"open" does not conform to the task requirements and
was replaced by the "openURI" method of the driver
base class in the updated task implementation.

"setDesiredBaudrate" has become redundant with the
addition of the "setSerialPortControlSettings"
and was only used in open.

Also adapted excecutables to use "openURI".
